(DOCSP-43059) Connect to MongoDB > Configure TLS (#51)
mongodb/docs-c: source/connect/tls.txt

.. c-tls:

========================================
Configure Transport Layer Security (TLS) 
========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: security, authentication, transport layer security, encrypt

Overview
--------

In this guide, you can learn how to use the :wikipedia:`TLS <Transport_Layer_Security>`
protocol to secure your connection to a MongoDB deployment.

When you enable TLS for a connection, the {+driver-short+} performs the following actions:

- Uses TLS to connect to the MongoDB deployment
- Verifies the deployment's certificate
- Ensures that the certificate certifies the deployment

To learn how to configure your MongoDB deployment for TLS, see the
:manual:`TLS configuration guide </tutorial/configure-ssl/>` in the
MongoDB Server manual. 

> .. important::

Reviewer: S: [nit] I feel like this admonition is more of a note than an important callout
```suggestion
.. note::
```

----------

You can enable TLS on a connection to your MongoDB instance
in the following ways:

- Setting the ``tls`` parameter in your connection string
- Using the ``mongoc_uri_set_option_as_bool()`` function to set the ``MONGOC_URI_TLS`` connection
  option to ``true``

.. tabs::

   .. tab:: Connection String
      :tabid: connectionstring

      .. literalinclude:: /includes/connect/tls-connection-str.c
         :language: c
         :start-after: start-connect-to-atlas
         :end-before: end-connect-to-atlas
         :dedent:

   .. tab:: MongoC URI Options
      :tabid: mongocurioptions

      .. literalinclude:: /includes/connect/tls-uri-options.c
         :language: c
         :start-after: start-connect
         :end-before: end-connect
         :dedent:

.. tip::
   
   If your connection string includes the ``+srv`` modification, which specifies the
   SRV connection format, TLS is enabled on your connection by default.

   To learn more about the SRV connection format, see
   :manual:`SRV Connection Format </reference/connection-string/#srv-connection-format>`
   in the {+mdb-server+} documentation.

.. _c-specify-ca-file:

Specify a CA File
------------------

> When MongoDB is started with TLS enabled, it will by default require the client to provide

Reviewer: S: since "MongoDB is started" is a little unclear, I'd reword to use active voice - maybe something like:
```suggestion
When you connect to a MongoDB deployment with TLS enabled, the deployment will by default require the client to provide
```

> a client certificate issued by a certificate authority specified by `--tlsCAFile`, or an authority

Reviewer: S: could you explain what specifying with `--tlsCAFile` means? or maybe just remove that part and make the wording a little more generic

```suggestion
a client certificate issued by a certificate authority, or an authority
```

trusted by the native certificate store in use on the server.

> To provide the client certificate, set the `tlsCertificateKeyFile` in the URI to a PEM armored

Reviewer: S: since the code example is using the `mongoc_uri_set_option_as_utf8()` function to set the `MONGOC_URI_TLSCERTIFICATEKEYFILE` option, rather than including `tlsCertificateKeyFile` in the URI, I'd adjust the wording to reflect that:
```suggestion
To provide the client certificate, use the `mongoc_uri_set_option_as_utf8()` function to set the MONGOC_URI_TLSCERTIFICATEKEYFILE option to to a `.pem` file containing the root certificate chain:
```

Reviewer: Is there also a way to set this option directly in the URI?

certificate file:

.. literalinclude:: /includes/connect/tls-ca-file.c
         :language: c
         :start-after: start-connect
         :end-before: end-connect
         :dedent:

.. _c-certificate-revocation:

Server Certificate Verification
-------------------------------

> The MongoDB C Driver will automatically verify the validity of the server certificate, such as issued
> by configured Certificate Authority, hostname validation, and expiration.

Reviewer: S: (if I'm understanding this correctly)

```suggestion
The MongoDB C Driver will automatically verify the validity of a server certificate issued
by the configured Certificate Authority. The driver also performs hostname validation and revocation checking.
```

To overwrite this behavior, it is possible to disable hostname validation, OCSP endpoint revocation checking,
revocation checking entirely, and allow invalid certificates.

This behavior is controlled using the ``tlsAllowInvalidHostnames``, ``tlsDisableOCSPEndpointCheck``,
``tlsDisableCertificateRevocationCheck``, and ``tlsAllowInvalidCertificates`` options. By default,
all are set to ``false``.

It is not recommended to change these defaults as it exposes the client to *Man In The Middle* attacks (when
``tlsAllowInvalidHostnames`` is set), invalid certificates (when ``tlsAllowInvalidCertificates`` is set), or
potentially revoked certificates (when ``tlsDisableOCSPEndpointCheck`` or ``tlsDisableCertificateRevocationCheck``
are set).

Supported Libraries
-------------------

By default, libmongoc will attempt to find a supported TLS library and enable TLS support. This is controlled

> by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. Valid values are:

Reviewer: S: clarify what these values are
```suggestion
by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. This flag accepts the following values:
```

> - ``AUTO`` the default behavior. Link to the system's native TLS library, or attempt to find OpenSSL.

Reviewer: S: Usually we use colons as [separators in lists](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items)

```suggestion
- ``AUTO``: Links to the system's native TLS library, or attempts to find OpenSSL. This is the default value.
```

> - ``DARWIN`` link to Secure Transport, the native TLS library on macOS.

Reviewer: S:
```suggestion
- ``DARWIN``: Links to Secure Transport, the native TLS library on macOS.
```

Reviewer: Applies to all list items

- ``WINDOWS`` link to Secure Channel, the native TLS library on Windows.
- ``OPENSSL`` link to OpenSSL (libssl). An optional install path may be specified with ``OPENSSL_ROOT``.
- ``LIBRESSL`` link to LibreSSL's libtls. (LibreSSL's compatible libssl may be linked to by setting ``OPENSSL``).
- ``OFF`` disable TLS support.

OpenSSL
~~~~~~~

The MongoDB C Driver uses OpenSSL on Linux and Unix platforms (besides macOS). Industry best practices
and some regulations require the use of TLS 1.1 or newer, which requires at least OpenSSL 1.0.1. Use the following command
to check your OpenSSL version:

.. code-block:: bash

>    $ openssl version

Reviewer: S: we try to avoid [using shell prompts](https://www.mongodb.com/docs/meta/style-guide/style/code-examples/#code-examples) ($) in the code:
```suggestion
   openssl version
```

Ensure your system's OpenSSL is a recent version (at least 1.0.1), or use the following command to install a recent version in a
non-system path and build against it:

.. code-block:: bash

   cmake -DOPENSSL_ROOT_DIR=/absolute/path/to/openssl

When compiled against OpenSSL, the driver will attempt to load the system default certificate store, as configured by the
distribution. That can be overridden by setting the ``tlsCAFile`` URI option or with the fields ``ca_file`` and ``ca_dir``
in the `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__.

> The Online Certificate Status Protocol (OCSP) (see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__) is fully supported
> when using OpenSSL 1.0.1+ with the following notes:
> 
> - When a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ and the ``crl_file`` revokes
>   the server's certificate, the certificate is considered revoked (even if the certificate has a valid stapled OCSP response).

Reviewer: S: S: I'd introduce that external link and avoid having information in parentheses. Also, since there's only one note in the list, you can just combine the info into one paragraph:

```suggestion
The Online Certificate Status Protocol (OCSP) is fully supported
when using OpenSSL 1.0.1+. However, when a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ and the ``crl_file`` revokes
  the server's certificate, the certificate is considered revoked, even if the certificate has a valid stapled OCSP response.
  
 .. tip::
 
   For more information about OCSP, see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__.
```

LibreSSL / libtls
~~~~~~~~~~~~~~~~~

The MongoDB C Driver supports LibreSSL through the use of OpenSSL compatibility checks when configured to compile against
``openssl``. It also supports the new ``libtls`` library when configured to build against ``libressl``.

When compiled against the Windows native libraries, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,
and will issue an error if used.

Setting ``tlsDisableOCSPEndpointCheck`` and ``tlsDisableCertificateRevocationCheck`` has no effect.

The Online Certificate Status Protocol (OCSP) (see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__) is partially supported
with the following notes:

> - The Must-Staple extension (see `RFC 7633 <https://tools.ietf.org/html/rfc7633>`__) is ignored.

Reviewer: S: same comment as above about introducing the external RFC links

- Connection will continue if a Must-Staple certificate is presented without a stapled response and the OCSP responder is down.
- Connection will not continue if the client receives a revoked response from an OCSP responder.

Native TLS Support on Windows (Secure Channel)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The MongoDB C Driver supports the Windows native TLS library (Secure Channel, or SChannel) and its native crypto library
(Cryptography API: Next Generation, or CNG).

When compiled against the Windows native libraries, the ``ca_dir`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported
and will issue an error if used.

> Encrypted PEM files (e.g., setting ``tlsCertificateKeyPassword``) are also not supported and will result in error when

Reviewer: S: can you explain what the options you mention in this section are? I'm guessing URI options
```suggestion
Encrypted PEM files (e.g., setting the ``tlsCertificateKeyPassword`` URI option) are also not supported and will result in error when
```

``System Local Machine Root`` certificate store.

Setting ``tlsDisableOCSPEndpointCheck`` has no effect.

The Online Certificate Status Protocol (OCSP) (see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__) is partially supported
with the following notes:

- The Must-Staple extension (see `RFC 7633 <https://tools.ietf.org/html/rfc7633>`__) is ignored.
- Connection will continue if a Must-Staple certificate is presented without a stapled response and the OCSP responder is down.
- Connection will not continue if the client receives a revoked response from an OCSP responder.
- When a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__, and the ``crl_file`` revokes the server's certificate, the OCSP response
  takes precedence. For example, if the server presents a certificate with a valid stapled OCSP response, the certificate is considered
  valid even if the ``crl_file`` marks it as revoked.

.. _secure-transport:

Native TLS Support on macOS / Darwin (Secure Transport)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The MongoDB C Driver supports the Darwin (OS X, macOS, iOS) native TLS library (Secure Transport), and its native crypto

Reviewer: S: I'm a little confused by the parentheses here; I think you can remove the "(OS X, macOS, iOS)" one, maybe explain briefly how "native TLS library relates" to "(Secure Transport)", and then just say "and Common Crypto, its native crypto library."

.. c-tls:

========================================

> Configure Transport Layer Security (TLS) 

Reviewer 2: Confirming: is https://www.mongodb.com/docs/languages/c/c-driver/current/libmongoc/guides/configuring_tls/ planned to be removed?

Author: yup – the docs are being standardized to the layout that other drivers docs follow, like the [pymongo docs](https://www.mongodb.com/docs/languages/kotlin/kotlin-sync-driver/current/indexes/), so this page will replace the previous 'configuring TLS' page

This behavior is controlled using the ``tlsAllowInvalidHostnames``, ``tlsDisableOCSPEndpointCheck``,
``tlsDisableCertificateRevocationCheck``, and ``tlsAllowInvalidCertificates`` options. By default,
all are set to ``false``.

It is not recommended to change these defaults, since you might expose your client to the following security risks:

- `Man In The Middle attacks <https://en.wikipedia.org/wiki/Man-in-the-middle_attack>`_, when
  ``tlsAllowInvalidHostnames`` is set
- Invalid certificates, when ``tlsAllowInvalidCertificates`` is set
- Potentially revoked certificates, when ``tlsDisableOCSPEndpointCheck`` or ``tlsDisableCertificateRevocationCheck`` are set

Supported Libraries
-------------------

By default, libmongoc will attempt to find a supported TLS library and enable TLS support. This is controlled
by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. This flag accepts the following values:

- ``AUTO``: Links to the system's native TLS library, or attempts to find OpenSSL. This is the default value.
- ``OPENSSL``: Links to OpenSSL (libssl). An optional install path may be specified with ``OPENSSL_ROOT``.

> - ``LIBRESSL``: Links to LibreSSL's libtls. You can link to LibreSSL's compatible libssl by setting ``OPENSSL``.

Reviewer 2: ```suggestion
- ``LIBRESSL`` (Deprecated):  Links to LibreSSL's libtls. You can link to LibreSSL's compatible libssl by setting ``OPENSSL``.
```

Suggest noting LIBRESSL is deprecated. [C driver 1.28.0](https://github.com/mongodb/mongo-c-driver/releases/tag/1.28.0) deprecates use of LibreSSL:

> `ENABLE_SSL=LIBRESSL` is deprecated along with `mongoc_stream_tls_libressl_new`. Support for LibreSSL may be dropped in a future major release.

to check your OpenSSL version:

.. code-block:: bash

   openssl version

Ensure your system's OpenSSL is a recent version (at least 1.0.1), or use the following command to install a recent version in a
non-system path and build against it:

.. code-block:: bash

   cmake -DOPENSSL_ROOT_DIR=/absolute/path/to/openssl

When compiled against OpenSSL, the driver will attempt to load the system default certificate store, as configured by the
distribution. That can be overridden by setting the ``tlsCAFile`` URI option or with the fields ``ca_file`` and ``ca_dir``
in the `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__.

The Online Certificate Status Protocol (OCSP) is fully supported
when using OpenSSL 1.0.1+. However, when a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ and the ``crl_file`` revokes
the server's certificate, the certificate is considered revoked, even if the certificate has a valid stapled OCSP response.
  
.. tip::
 
   For more information about OCSP, see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__.

> LibreSSL / libtls
> ~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
LibreSSL / libtls (Deprecated)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The MongoDB C Driver supports LibreSSL through the use of OpenSSL compatibility checks when configured to compile against
``openssl``. It also supports the new ``libtls`` library when configured to build against ``libressl``.

> When compiled against the Windows native libraries, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,

Reviewer 2: ```suggestion
When compiled with LibreSSL, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,
```

      :tabid: connectionstring

      .. literalinclude:: /includes/connect/tls.c
         :language: c
         :start-after: start-connect-str
         :end-before: end-connect-str
         :dedent:

   .. tab:: MongoC URI Options
      :tabid: mongocurioptions

      .. literalinclude:: /includes/connect/tls.c
         :language: c
         :start-after: start-connect-uri-opt
         :end-before: end-connect-uri-opt
         :dedent:

.. tip::
   
   If your connection string includes the ``+srv`` modification, which specifies the
   SRV connection format, TLS is enabled on your connection by default.

   To learn more about the SRV connection format, see
   :manual:`SRV Connection Format </reference/connection-string/#srv-connection-format>`
   in the {+mdb-server+} documentation.

.. _c-specify-ca-file:

Specify a CA File
------------------

When you connect to a MongoDB deployment with TLS enabled, the deployment will by default require the client to provide
a client certificate issued by a certificate authority, or an authority
trusted by the native certificate store in use on the server.

You can provide the client certificate in the following ways:

> - Setting the ``tlscertificatekeyfile`` parameter in your connection string to a PEM armored certificate file

Reviewer: S: I'd use the same description of a pem file in both these bullet points (and use the same placeholder value in the code tabs)
```suggestion
- Setting the ``tlscertificatekeyfile`` parameter in your connection string to a ``.pem`` file containing the root certificate chain
```

- Using the `mongoc_uri_set_option_as_utf8()` function to set the `MONGOC_URI_TLSCERTIFICATEKEYFILE` option
  to a `.pem` file containing the root certificate chain

.. tabs::

   .. tab:: Connection String
      :tabid: connectionstring

      .. literalinclude:: /includes/connect/tls.c
         :language: c
         :start-after: start-connect-ca-file-str
         :end-before: end-connect-ca-file-str
         :dedent:

   .. tab:: MongoC URI Options
      :tabid: mongocurioptions

      .. literalinclude:: /includes/connect/tls.c
         :language: c
         :start-after: start-connect-ca-file-uri
         :end-before: end-connect-ca-file-uri
         :dedent:

.. _c-certificate-revocation:

Server Certificate Verification
-------------------------------

The MongoDB C Driver will automatically verify the validity of a server certificate issued
by the configured Certificate Authority. The driver also performs hostname validation and revocation checking.

To overwrite this behavior, it is possible to disable hostname validation, OCSP endpoint revocation checking,

> revocation checking entirely, and allow invalid certificates.

Reviewer: S:
```suggestion
all revocation checking, and allow invalid certificates.
```

This behavior is controlled using the ``tlsAllowInvalidHostnames``, ``tlsDisableOCSPEndpointCheck``,
``tlsDisableCertificateRevocationCheck``, and ``tlsAllowInvalidCertificates`` options. By default,
all are set to ``false``.

> It is not recommended to change these defaults as it exposes the client to *Man In The Middle* attacks (when

Reviewer: S: maybe use an inline link to a wikipedia page describing man in the middle attacks instead of italics, since readers might not know what that is

> It is not recommended to change these defaults as it exposes the client to *Man In The Middle* attacks (when

Reviewer: S: I think this paragraph would be more readable as a list; you could also avoid parentheses that way. Something like: "It is not recommended to change these defaults, since you might expose your client to the following security risks:
- Man in the middle attacks, when ``tlsAllowInvalidHostnames`` is set
...

``tlsAllowInvalidHostnames`` is set), invalid certificates (when ``tlsAllowInvalidCertificates`` is set), or
potentially revoked certificates (when ``tlsDisableOCSPEndpointCheck`` or ``tlsDisableCertificateRevocationCheck``
are set).

Supported Libraries
-------------------

By default, libmongoc will attempt to find a supported TLS library and enable TLS support. This is controlled
by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. This flag accepts the following values:

> 

Reviewer: S: [nit] I'd list these values in the same order as the subsections appear

- ``AUTO``: Links to the system's native TLS library, or attempts to find OpenSSL. This is the default value.
- ``DARWIN``: Links to Secure Transport, the native TLS library on macOS.
- ``WINDOWS``: Links to Secure Channel, the native TLS library on Windows.
- ``OPENSSL``: Links to OpenSSL (libssl). An optional install path may be specified with ``OPENSSL_ROOT``.

> - ``LIBRESSL``: Links to LibreSSL's libtls. (LibreSSL's compatible libssl may be linked to by setting ``OPENSSL``).

Reviewer: S: I'd reword to active voice and remove the parentheses
```suggestion
- ``LIBRESSL``: Links to LibreSSL's libtls. You can link to LibreSSL's compatible libssl by setting ``OPENSSL``.
```

distribution. That can be overridden by setting the ``tlsCAFile`` URI option or with the fields ``ca_file`` and ``ca_dir``
in the `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__.

The Online Certificate Status Protocol (OCSP) is fully supported
when using OpenSSL 1.0.1+. However, when a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ and the ``crl_file`` revokes
the server's certificate, the certificate is considered revoked, even if the certificate has a valid stapled OCSP response.
  
.. tip::
 
   For more information about OCSP, see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__.

LibreSSL / libtls
~~~~~~~~~~~~~~~~~

The MongoDB C Driver supports LibreSSL through the use of OpenSSL compatibility checks when configured to compile against
``openssl``. It also supports the new ``libtls`` library when configured to build against ``libressl``.

When compiled against the Windows native libraries, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,
and will issue an error if used.

> Setting ``tlsDisableOCSPEndpointCheck`` and ``tlsDisableCertificateRevocationCheck`` has no effect.

Reviewer: Q: setting these options has no effect when compliling against the Windows native libraries, or always? If it's the former, I'd combine this sentence with the paragraph before

The Online Certificate Status Protocol (OCSP) is partially supported
with the following notes:

- The Must-Staple extension (see `RFC 7633 <https://tools.ietf.org/html/rfc7633>`__) is ignored.
- Connection will continue if a Must-Staple certificate is presented without a stapled response and the OCSP responder is down.
- Connection will not continue if the client receives a revoked response from an OCSP responder.

.. tip::
 
   For more information about OCSP, see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__.

Native TLS Support on Windows (Secure Channel)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The MongoDB C Driver supports the Windows native TLS library (Secure Channel, or SChannel) and its native crypto library
(Cryptography API: Next Generation, or CNG).

When compiled against the Windows native libraries, the ``ca_dir`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported
and will issue an error if used.

> Encrypted PEM files (e.g., setting the ``tlsCertificateKeyPassword`` URI option) are also not supported and will result in error when

Reviewer: S: avoiding [parentheses](https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#parentheses):
```suggestion
Encrypted PEM files, set by using the ``tlsCertificateKeyPassword`` URI option, are also not supported and will result in error when
```

attempting to load them.

> When ``tlsCAFile`` is set, the driver will only allow server certificates issued by the authority (or authorities) provided.

Reviewer: S: parentheses
```suggestion
When ``tlsCAFile`` is set, the driver will only allow server certificates issued by one or more authorities provided.
```

- When a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__, and the ``crl_file`` revokes the server's certificate, the OCSP response
  takes precedence. For example, if the server presents a certificate with a valid stapled OCSP response, the certificate is considered
  valid even if the ``crl_file`` marks it as revoked.

.. tip::
 
   For more information about OCSP, see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__.

.. _secure-transport:

Native TLS Support on macOS / Darwin (Secure Transport)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The MongoDB C Driver supports both the Darwin native TLS library and Common Crypto, its native crypto
library.

When compiled against Secure Transport, the ``ca_dir`` and ``crl_file`` options of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ are not supported.
An error is issued if either are used.

When ``tlsCAFile`` is set, the driver will only allow server certificates issued by the authority (or authorities) provided.
When no ``tlsCAFile`` is set, the driver will use the Certificate Authorities in the unlocked keychains.

> Setting ``tlsDisableOCSPEndpointCheck`` and ``tlsDisableCertificateRevocationCheck`` has no effect.

Reviewer 2: When compiling against secure transport ([checked here](https://github.com/mongodb/mongo-c-driver/blob/56cbe883be2fe84812e35ea8fae181dc7e0be8ce/src/libmongoc/src/mongoc/mongoc-stream-tls-secure-transport.c#L695-L698)).
